python/caffe/_caffe.cpp

```diff
@@ -507,13 +508,15 @@ BOOST_PYTHON_MODULE(_caffe) {
     .def("restore", &Solver<Dtype>::Restore)
     .def("snapshot", &Solver<Dtype>::Snapshot)
     .def("share_weights", &share_weights)
+    .def("apply_update", &Solver<Dtype>::ApplyUpdate)
     .add_property("param", bp::make_function(&Solver<Dtype>::param,
-              bp::return_value_policy<bp::copy_const_reference>()));
+              bp::return_internal_reference<>()));
   BP_REGISTER_SHARED_PTR_TO_PYTHON(Solver<Dtype>);
 
   bp::class_<SGDSolver<Dtype>, bp::bases<Solver<Dtype> >,
     shared_ptr<SGDSolver<Dtype> >, boost::noncopyable>(
-        "SGDSolver", bp::init<string>());
+        "SGDSolver", bp::init<string>())
+        .def("get_learning_rate", &SGDSolver<Dtype>::GetLearningRate);
```

Comment:
What about other solvers, will they get this property only due to the inheritance?

Reply:
Other solvers extending from `SGDSolver`? I think only them will get this, but this is OK, no? this is on purpose.

Reply:
So if I do `a = caffe.AdamSolver("solver.prototxt")` in my pycaffe script, will I be able to call `a.get_learning_rate()` or not?

Reply:
To my understanding, the answer is yes, but I have not tested it.

Reply:
Would be good to know this. Maybe consider implementing some [tests](https://github.com/BVLC/caffe/blob/master/python/caffe/test/test_solver.py) for this change? See #4342, it is somewhat similar to this (a bit more clumsy) but also had some very simple tests.

Reply:
No, this does not inherit correctly b.c. the `bp::bases` of the child `SGDSolver`s are set to `Solver` and not `SGDSolver`. Switching the `NesteroveSolver` and other children to have the `SGDSolver` base fixes the issue. I second incorporating the `apply_update()` test from c5b5b55845c20930026f1f563136247c41f58075, so could you cherry-pick it to give @nitnelave credit for it?